tests/Cache/FileSystemCacheItemPoolTest.php

```diff
@@ -0,0 +1,229 @@
+        ];
+
+        foreach ($items as $item) {
+            $this->pool->save($item);
+        }
+
+        $keys = array_map(function (CacheItemInterface $item) {
+            return $item->getKey();
+        }, $items);
+        array_push($keys, 'NonExistant');
+
+        $retrievedItems = $this->pool->getItems($keys);
+
+        foreach ($items as $item) {
+            $this->assertTrue($retrievedItems[$item->getKey()]->isHit());
+        }
+
+        $this->assertFalse($retrievedItems['NonExistant']->isHit());
+    }
+
+    public function testClear()
+    {
+        $item = $this->getNewItem();
+        $this->pool->save($item);
+        $this->assertLessThan(scandir($this->defaultCacheDirectory), 2);
+        $this->pool->clear();
+        // Clear removes all the files, but scandir returns `.` and `..` as files
+        $this->assertEquals(count(scandir($this->defaultCacheDirectory)), 2);
+    }
+
+    public function testSaveDeferredAndCommit()
+    {
+        $item = $this->getNewItem();
+        $this->pool->saveDeferred($item);
+        $this->assertFalse($this->pool->getItem($item->getKey())->isHit());
+
+        $this->pool->commit();
+        $this->assertTrue($this->pool->getItem($item->getKey())->isHit());
+    }
+
+    public function testGetItemWithIncorrectKeyShouldThrowAnException()
+    {
+        foreach ($this->invalidChars as $char) {
+            try {
+                $item = $this->getNewItem($char);
+                $this->pool->getItem($item->getKey());
+                $this->fail('The key ' . $char . ' is passing validation when should not be valid');
+            } catch (InvalidArgumentException $e) {
+                $this->assertTrue($e instanceof InvalidArgumentException);
+            }
+        }
+    }
+
+    public function testGetItemsWithIncorrectKeyShouldThrowAnException()
+    {
+        foreach ($this->invalidChars as $char) {
+            try {
+                $item = $this->getNewItem($char);
+                $this->pool->getItems([$item->getKey()]);
+                $this->fail('The key ' . $char . ' is passing validation when should not be valid');
+            } catch (InvalidArgumentException $e) {
+                $this->assertTrue($e instanceof InvalidArgumentException);
+            }
+        }
+    }
+
+    public function testHasItemWithIncorrectKeyShouldThrowAnException()
+    {
+        foreach ($this->invalidChars as $char) {
+            try {
+                $item = $this->getNewItem($char);
+                $this->pool->hasItem($item->getKey());
+                $this->fail('The key ' . $char . ' is passing validation when should not be valid');
+            } catch (InvalidArgumentException $e) {
+                $this->assertTrue($e instanceof InvalidArgumentException);
+            }
+        }
+    }
+
+    public function testDeleteItemWithIncorrectKeyShouldThrowAnException()
+    {
+        foreach ($this->invalidChars as $char) {
```

Comment:
instead of using a `foreach` with `invalidChars`, use a `@dataProvider` annotation. For example:

```php
    /**
     * @dataProvider provideInvalidChars
     */
    public function testDeleteItemWithIncorrectKeyShouldThrowAnException($char)
    {
        $this->expectException(InvalidArgumentException::class);
        $this->expectExceptionMessage('Character ' . $char . ' is invalid'); // or whatever the exception message really is
        $item = $this->getNewItem($char);
        $this->pool->hasItem($item->getKey());
    }

    public function provideInvalidChars()
    {
        // each element in the array is an array of arguments, so wrap each char in an array
        return array_map(fn($char) => [$char], $this->invalidChars);
    }
```